### PR TITLE
UtilPrng handle Implementation

### LINKED ISSUE #5724: Modularized Implementation of Prng Service
In phase one, we will see a full CI run pass, with PRNG Service operations largely performed by workflows and modularized business logic and the Merkle tree root being `MerkleHederaState`.

- [x] Re-do the following in PBJ branch (ensure HCS e2e tests pass again): 
   - payer sig
   - solvency check
   - throttle APIs
   - `HederaMerkleState.preHandle()` to choose between `mono` and `app` implementations based on operation type
   - signature expansion

### INGEST
Add `UtilPrng` to the `workflows.enabled` property, binding both `mono-service` and `hedera-app` gRPC services as in the Consensus Service. Right now the server only binds the services for Consensus Service. Update the code that will bind the mono-service PRNG functionality and workflows UtilPrng functionality (see #4947 for an example)

### Client E2E:
There will be a matching property workflow.operations=... on the test clients side to recognize which port to use for which HAPI operation.

All Prng Service operations will be included in workflow.operations.
All Prng Service-specific prechecks will done by a modularized Handler, using state obtained via the States API.
Any additional facility the Prng Service handler needs MAY be implemented by a mono-service adapter.
The ingest workflow MAY implement any infrastructure it needs via an adapter.

_Tasks_
 - [x] Update `ServicesMain` to bind PRNG functionalities for both `mono-service` and `hedera-app`
 - [ ] Update EET mechanism with Prng `client.workflow.operations` to differentiate between `mono-service` Prng functionality and workflows Prng functionality

### PRE-HANDLE 
Pre-handle will now be "workflow-aware", so within `HederaMerkleState#preHandle`, we will first extract the function, and then either call into the `mono-service`signature expansion; or into the pre-handle workflow.

_Tasks_
- [x] Enhance `NetworkService` to retrieve N-Minus-3-Running-Hash
- [x] Implement a hand-off of Network Service's hash to `PrngService`
- [x] Implement `UtilPrngHandler.preHandle()`

### HANDLE 
Once again handle will become workflow-aware, so that within `HederaMerkleState` we can immediately adapt a platform `Transaction` with metadata derived from the pre-handle workflow into a `SignedTxnAccessor`. 

However, instead of delegating the final business logic to a `TransitionLogic`, this must be done via a `Handler` that (as in INGEST),
 - Uses the States API for persistence.
 - Receives facilities/context with API's that approximate the final SPI versions.

_Tasks_
- [x] Implement `UtilPrngHandler.handle()`
- [x] Enhance minimal record-builder API to capture the information from `UtilPrngHandler`
- [x] Use Config Facility Adapter to provide config needed (for enabling/disabling Prng functionality)

### PBJ
 - [x] Depends on PBJ changes (#4712)

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes #5724 

- Implements `UtilPrngHandler`, `handle()` and `prehandle()` methods
- Added [createStore](https://github.com/hashgraph/hedera-services/blob/5724-D-util-prng-handle/hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/MonoHandleContext.java#L96) to `MonoHandleContext`  as a temporary solution, until `HandleContext` is in place
- To inject `ReadableStoreFactory` to `MonoHandleContext` provided it in [HandleWorkflowModule](https://github.com/hashgraph/hedera-services/blob/5724-D-util-prng-handle/hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflowModule.java#L94)
- Added [ReadableRunningHashLeafStore](https://github.com/hashgraph/hedera-services/blob/5724-D-util-prng-handle/hedera-node/hedera-network-service/src/main/java/com/hedera/node/app/service/network/ReadableRunningHashLeafStore.java) in `NetworkService` to provide `nMinus3RunningHash`
- [PrngConfig](https://github.com/hashgraph/hedera-services/blob/5724-D-util-prng-handle/hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/config/PrngConfig.java) is added to validate if the feature is enabled. But this will move to some workflow in the future

Testing:
- Unit tests
- Did not test with `workflows.enabled`. As far as I know, we can't test it until some other changes to workflows are merged

### CODE REVIEW

File: hedera-node/hapi-fees/src/main/java/module-info.java [hunk 1]
```diff
@@ -22,4 +22,5 @@
     requires com.google.protobuf;
     requires org.apache.commons.lang3;
     requires com.google.common;
+    requires com.hedera.hashgraph.protobuf.java.api;
```
Comment (Neeharika-Sompalli) [inline hedera-node/hapi-fees/src/main/java/module-info.java:25]: This is added because IntelliJ showed errors, but the gradle build still succeeded

File: hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import com.hedera.node.app.service.network.ReadableRunningHashLeafStore;
+import com.hedera.node.app.spi.state.ReadableSingletonState;
+import com.hedera.node.app.spi.state.ReadableStates;
+import com.swirlds.common.crypto.RunningHash;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * The underlying data storage class that holds the record running hash leaf data.
+ */
+public class ReadableRunningHashLeafStoreImpl implements ReadableRunningHashLeafStore {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java:32]: I believe the app will need to expose the `n-3` running hash through a custom interface, since the app will be managing the record stream in state as a special Merkle leaf outside the `States` API.

So although this is quite reasonable for this PR, I would slightly prefer---as a first approximation---to add a `RunningHash getTrailingHash(int n)` method on the `HandleContext`.

(Note the `ContractService` will **also** need the `n-1` trailing hash, both for its independent cache of block hashes and to implement `blockhash(block.number - 1)` in a transaction that is first in a block.)

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+package com.hedera.node.app.spi.fixtures;
+
+import java.util.UUID;
+
+public class Utils {
+    public static byte[] randomUtf8Bytes(int n) {
+        byte[] data = new byte[n];
+        int i = 0;
+        while (i < n) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java:25]: Curious, why did we write this instead of using SecureRandom?
```
SecureRandom.getInstanceStrong().nextBytes(data);
```
Is SecureRandom not random enough?

File: hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java:32]: Wouldn't NetworkService be responsible for the record running hash leaf ?

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java (hunk 3 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java:25]: No, I think I can use `SecureRandom`. This was a method in `mono-service`. I re-used it and moved to testFixtures.

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java [hunk 4]
```diff
@@ -18,27 +18,139 @@
...
+        try {
+            nMinusThreeHash = nMinus3RunningHash.getFutureHash().get();
+            // Use n-3 running hash instead of n-1 running hash for processing transactions quickly
+            if (nMinusThreeHash == null || Arrays.equals(nMinusThreeHash.getValue(), new byte[48])) {
+                log.info("No n-3 record running hash available to generate random number");
+                return MISSING_BYTES;
+            }
+            // generate binary string from the running hash of records
+            return nMinusThreeHash.getValue();
+        } catch (InterruptedException | ExecutionException e) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:137]: We probably shouldn't try to combine these two.  Interrupted means the process is being asked to exit "cleanly".  ExecutionException is just any exception (runtime, checked, etc..) that occurred on the other thread.  The first should interrupt, but I don't thing ExecutionException should do that.
Also looks like a partial mingle at the bottom with a TODO after throwing an exception.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/HandleContext.java [hunk 5]
```diff
@@ -58,4 +59,16 @@ public interface HandleContext {
...
+     * Create a new store given the store's interface. This gives read-only access to the store.
+     *
+     * @param storeInterface The store interface to find and create a store for
+     * @return An implementation of store interface provided, or null if the store
+     * @param <C> Interface class for a Store
+     * @throws IllegalArgumentException if the storeInterface class provided is unknown to the app
+     * @throws NullPointerException if {@code clazz} is {@code null}
+     */
+    @NonNull
+    <C> C createStore(@NonNull Class<C> storeInterface);
```
Comment (netopyr) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/HandleContext.java:73]: I think we should name this method `createReadableStore()` to avoid any confusion with writable stores.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java [hunk 6]
```diff
@@ -62,6 +68,7 @@ ReadableScheduleStore.class, new StoreEntry(ScheduleService.NAME, ReadableSchedu
      *
      * @param state the {@link HederaState} to use
      */
+    @Inject
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java:71]: A `ReadableStoreFactory` should not be injected.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/MonoHandleContext.java [hunk 7]
```diff
@@ -39,17 +40,20 @@ public class MonoHandleContext implements HandleContext {
...
+    private final ReadableStoreFactory readableStoreFactory;
 
     @Inject
     public MonoHandleContext(
             @NonNull final EntityIdSource ids,
             @NonNull final ExpiryValidator expiryValidator,
             @NonNull final AttributeValidator attributeValidator,
-            @NonNull final TransactionContext txnCtx) {
+            @NonNull final TransactionContext txnCtx,
+            @NonNull final ReadableStoreFactory storeFactory) {
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/MonoHandleContext.java:51]: Are you sure, this is correct? `MonoHandleContext` has singleton-scope while a `ReadableStoreFactory` is bound to a particular state.

In the upcoming design, which will be similar to what we have in pre-handle already, a context will be created per transaction and it makes sense to add `ReadableStoreFactory`. But here, I think it is wrong.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflowModule.java [hunk 8]
```diff
@@ -87,4 +88,9 @@ static Supplier<AutoCloseableWrapper<HederaState>> provideStateSupplier(@NonNull
     static LongSupplier provideConsensusSecond(@NonNull final TransactionContext txnCtx) {
         return () -> txnCtx.consensusTime().getEpochSecond();
     }
+
+    @Provides
+    static ReadableStoreFactory provideReadableStoreFactory(final Supplier<AutoCloseableWrapper<HederaState>> state) {
+        return new ReadableStoreFactory(state.get().get());
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflowModule.java:94]: Which state is used when a `ReadableStoreFactory` is created like this?

File: hedera-node/hedera-app/src/main/java/module-info.java [hunk 9]
```diff
@@ -61,4 +61,5 @@
             com.swirlds.platform;
     exports com.hedera.node.app.throttle to
             com.swirlds.platform;
+    exports com.hedera.node.app.workflows.dispatcher;
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/module-info.java:64]: We should not export anything from the dispatcher package IMO. Why is this needed?

File: hedera-node/hedera-network-service/src/main/java/com/hedera/node/app/service/network/ReadableRunningHashLeafStore.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Provides access to the underlying running hash leaf in state.
+ */
+public interface ReadableRunningHashLeafStore {
+    /**
+     * Get the n-3 record's running hash from running hash leaf.
+     * @return the n-3 record's running hash
+     */
+    RunningHash getNMinusThreeRunningHash();
```
Comment (netopyr) [inline hedera-node/hedera-network-service/src/main/java/com/hedera/node/app/service/network/ReadableRunningHashLeafStore.java:29]: `NonNull` annotation missing

File: hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public ReadableRunningHashLeafStoreImpl(@NonNull final ReadableStates states) {
+        requireNonNull(states);
+        this.runningHashState = states.getSingleton(RUNNING_HASHES_KEY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public RunningHash getNMinusThreeRunningHash() {
```
Comment (netopyr) [inline hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java:49]: `Override` annotation missing

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/config/PrngConfig.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.util.impl.config;
+
+/**
+ * Configuration for the PRNG to check if it is enabled or not.
+ * @param isPrngEnabled true if the PRNG is enabled, false otherwise
+ */
+public record PrngConfig(boolean isPrngEnabled) {}
```
Comment (netopyr) [inline hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/config/PrngConfig.java:23]: I think it is kind of unusual to name a property `isSomething`. Verbs are usually reserved for methods. Is this something we also do elsewhere in the services-repository?

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java [hunk 13]
```diff
@@ -18,27 +18,139 @@
...
 public class UtilPrngHandler implements TransactionHandler {
+    private static final Logger log = LogManager.getLogger(UtilPrngHandler.class);
+    public static final byte[] MISSING_BYTES = new byte[0];
+
     @Inject
     public UtilPrngHandler() {
         // Dagger2
     }
 
-    @Override
```
Comment (netopyr) [inline hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:39]: Please keep the `Override` annotation. It ensures that compiler and developer are immediately in the know.

(An `{@inheritDoc}` by itself on the other hand is redundant IMO, in particular javadoc does not need it. But I noticed some people insist on having it. Guess, it does not hurt either.)

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java [hunk 14]
```diff
@@ -18,27 +18,139 @@
...
+        }
+    }
+
+    /**
+     * Generate a random number from the given bytes in the given range
+     * @param pseudoRandomBytes bytes to generate random number from
+     * @param range range of the random number
+     * @return random number
+     */
+    private int randomNumFromBytes(@NonNull final byte[] pseudoRandomBytes, final int range) {
```
Comment (netopyr) [inline hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:152]: Not sure how crucial real randomness is in this context, but I think depending on the `range` this function will return slightly skewed results. Is this acceptable?

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java [hunk 15]
```diff
@@ -18,27 +18,139 @@
...
+        final var initialBitsValue = ByteBuffer.wrap(pseudoRandomBytes, 0, 4).getInt();
+        return LongMath.mod(initialBitsValue, range);
```
Comment (netopyr) [inline hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:154]: `initialBitsValue` is an `int`, right? What is the advantage of using `LongMath` here?

File: hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/records/UtilPrngRecordBuilderTest.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    private UtilPrngRecordBuilder subject;
+
+    @BeforeEach
+    void setUp() {
+        subject = new UtilPrngRecordBuilder();
+    }
+
+    @Test
+    void emptyConstructor() {
+        assertNull(subject.getPrngBytes());
```
Comment (netopyr) [inline hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/records/UtilPrngRecordBuilderTest.java:40]: I believe we agreed to use AssertJ in new tests.

File: hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/handlers/UtilPrngHandlerTest.java [hunk 17]
```diff
@@ -30,10 +48,215 @@ class UtilPrngHandlerTest {
...
+        givenTxnWithoutRange();
+    }
+
+    @Test
+    void preHandleValidatesRange() {
+        final var body = TransactionBody.newBuilder()
+                .utilPrng(UtilPrngTransactionBody.newBuilder())
+                .build();
+        given(preHandleContext.body()).willReturn(body);
+        assertDoesNotThrow(() -> subject.preHandle(preHandleContext));
```
Comment (netopyr) [inline hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/handlers/UtilPrngHandlerTest.java:78]: I believe we agreed to use AssertJ in new tests.

**Review by netopyr — CHANGES_REQUESTED** [review]
I think we should discuss how `ReadableStoreFactory` needs to be integrated. I am not sure, the current approach works.

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.spi.fixtures;
+
+import java.util.UUID;
+
+public class Utils {
+    public static byte[] randomUtf8Bytes(int n) {
```
[the remark this answers is not in the document]
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java:22]: I agree, we should use pseudorandom with a seed that will always return the same random values from test run to test run.

File: hedera-node/hapi-fees/src/main/java/module-info.java (hunk 1 above)
Comment (rbair23) [reply hedera-node/hapi-fees/src/main/java/module-info.java:25]: I wonder if you removed this line and rebuilt the intellij caches if the error would go away?

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java (hunk 3 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java:25]: Just saw that there is a randomBytes method in TestBase, will use that.

File: hedera-node/hapi-fees/src/main/java/module-info.java (hunk 1 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hapi-fees/src/main/java/module-info.java:25]: No it still shows errors. Not sure if its my IntelliJ. But the `./gradlew assemble` passes fine.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflowModule.java (hunk 8 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflowModule.java:94]: Since we are passing in `Supplier<AutoCloseableWrapper<HederaState>>` it uses `platform.getLatestImmutableState` which is provided as 
```
  @Provides
    @SuppressWarnings({"unchecked", "rawtypes"})
    static Supplier<AutoCloseableWrapper<HederaState>> provideStateSupplier(@NonNull final Platform platform) {
        // Always return the latest immutable state until we support state proofs
        return () -> (AutoCloseableWrapper) platform.getLatestImmutableState(HandleWorkflowModule.class.getName());
    }
```

File: hedera-node/hedera-network-service/src/main/java/com/hedera/node/app/service/network/ReadableRunningHashLeafStore.java (hunk 10 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-network-service/src/main/java/com/hedera/node/app/service/network/ReadableRunningHashLeafStore.java:29]: Added. Thanks

File: hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java (hunk 11 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-network-service-impl/src/main/java/com/hedera/node/app/service/network/impl/ReadableRunningHashLeafStoreImpl.java:49]: Added. Thanks

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/config/PrngConfig.java (hunk 12 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/config/PrngConfig.java:23]: I named it that way based on `utilPrng.isEnabled`. WIll rename to `prngEnabled`

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java (hunk 14 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:152]: Yeah, since we are not changing the business logic and it is how the HIP specified. I think it is fine

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java (hunk 13 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:39]: Yes, removed by mistake. Put it back

File: hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/records/UtilPrngRecordBuilderTest.java (hunk 16 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/records/UtilPrngRecordBuilderTest.java:40]: Changing them

File: hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/handlers/UtilPrngHandlerTest.java (hunk 17 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/handlers/UtilPrngHandlerTest.java:78]: Changed to use AssertJ

File: hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/records/UtilPrngRecordBuilderTest.java (hunk 16 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/test/java/com/hedera/node/app/service/util/impl/test/records/UtilPrngRecordBuilderTest.java:40]: Changed to use AssertJ

File: hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/Utils.java:22]: Changed to use `TestBase` randomBytes method

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java (hunk 6 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java:71]: Not sure how else to provide to `MonoHandleContext`. Do you think it should be provided in `createReadableStore` ?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/MonoHandleContext.java (hunk 7 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/MonoHandleContext.java:51]: Yeah, I have added as temporary solution, until `HandleContext` is added to be used in `HandleWorkflow`.
Do you have any other ideas how can I inject it for `createReadableStore` ?

Should I inject `ReadableStoreFactory`  in `UtilPrng.handle()` method and provide it to `createReadableStore` ? @netopyr

File: hedera-node/hedera-app/src/main/java/module-info.java (hunk 9 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/module-info.java:64]: Yeah because I used ReadableStoreFactory in `MonoHandleContext`.

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java (hunk 15 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:155]: yeah I am not sure. We were using `LongMath.mod` in `mono-service`. I think it can be changed to `IntMath.mod`. Changed it.

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java (hunk 4 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java]: Separated both exceptions. Added `UNKNOWN` response, which might be changed in the future

File: hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java [hunk 19]
```diff
@@ -18,27 +18,144 @@
...
+
+    /**
+     * Generate a random number from the given bytes in the given range
+     * @param pseudoRandomBytes bytes to generate random number from
+     * @param range range of the random number
+     * @return random number
+     */
+    private int randomNumFromBytes(@NonNull final byte[] pseudoRandomBytes, final int range) {
+        final int initialBitsValue = ByteBuffer.wrap(pseudoRandomBytes, 0, 4).getInt();
+        return IntMath.mod(initialBitsValue, range);
```
Comment (jsync-swirlds) [inline hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:159]: Is there a reason we used Google's IntMath class instead of the modulus operator (`%`)?

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:159]: If I remember correctly,  it was suggested by Leemon when the HIP was implemented.

Comment (jsync-swirlds) [reply hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java:159]: It's extra dependency and slower than the operator, but probably doesn't matter in this method.  The biggest difference is that Google's method would throw ArithmeticException if range was negative and flip the result if initialBitsValue was negative.  Internally it still calls the operator.

**Review by jsync-swirlds — DISMISSED** [review]

**Review by jsync-swirlds — DISMISSED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java (hunk 6 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java:71]: Added Dagger injected State and Provider<StoreFactory> , so it will get latest mutable state at start of `onHandleConsensusRound` , and will create new factory for each handle. This is added as temporary solution until `HandleContextImpl` is implemented.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/HandleScope.java [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.lang.annotation.Retention;
+import javax.inject.Scope;
+
+/**
+ * A Dagger scope that indicates that a component's lifetime is for each handle.
+ */
+@Scope
+@Retention(RUNTIME)
+public @interface HandleScope {}
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/HandleScope.java:29]: hmmm.  Is the scope properly to Handle (kind of implies each method), or each transaction?  Would Transaction (i.e. grpc `@CallScope`) be clearer here?

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/meta/HandleScope.java:29]: Yeah it per each `handle` call which happens for each transaction. @CallScope, might be ambiguous on which call

**Review by jsync-swirlds — DISMISSED** [review]
Needs rebase, but LGTM.

**Review by netopyr — APPROVED** [review]
LGTM

**Review by jsync-swirlds — APPROVED** [review]

**Review by qnswirlds — APPROVED** [review]

**Review by povolev15 — APPROVED** [review]
